controllers/controllers.go

```diff
@@ -281,3 +296,55 @@ func markAsFreeOnSuccess(c client.Client, watchNamespace string, recorder record
 	}
 	return err
 }
+
+// markNodeAsUpgrading marks the given node as upgrading by adding an annotation to it. If the number of nodes
+// performing upgrades in parallel exceeds the maximum allowed, an error is returned.
+func markNodeAsUpgrading(ctx context.Context, c client.Client, node *core.Node) error {
+	controllerLocker.Lock()
+	defer controllerLocker.Unlock()
+
+	upgradingNodes, err := findUpgradingNodes(ctx, c, node)
+	if err != nil {
+		return err
+	}
+	if len(upgradingNodes) == 1 && upgradingNodes[0].Name == node.Name {
+		// node is already marked as upgrading
+		return nil
+	}
+
+	// check if marking the node as upgrading would exceed the maximum allowed
+	if len(upgradingNodes) >= MaxParallelUpgrades {
+		return fmt.Errorf("cannot mark node %s as upgrading, maximum number of parrallel upgrading nodes reached (%d)",
+			node.Name, MaxParallelUpgrades)
+	}
+	if err := metadata.ApplyUpgradingAnnotation(ctx, c, *node); err != nil {
+		return err
+	}
+	return nil
+}
+
+// findUpgradingNodes returns a list of nodes that are currently upgrading i.e. have the upgrading annotation set to
+// true. As shortcut-circuit if the given node is already marked as upgrading, it is returned as the only element
+// in the list.
+func findUpgradingNodes(ctx context.Context, c client.Client, node *core.Node) ([]core.Node, error) {
```

Comment:
This function is doing too much. You can tell as the doc is having to explain behavior that contradicts the name `findUpgradingNodes`.

Please change this to just return a list of upgrading nodes, the caller can then determine if the node is in the list, and handle that correct.

Reply:
In fact this whole function can be removed if a label was used instead of an annotation. [Labels should be used when the use case is to find collections of objects that satisfy certain conditions](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#attaching-metadata-to-objects). Which is the case here.

That simplifies this to:
```
	if err := c.List(ctx, nodeList, client.MatchingLabels{core.LabelOSStable: "windows", metadata.UpgradingAnnotation: "true"}); err != nil {
		return nil, fmt.Errorf("error listing Windows nodes: %w", err)
	}
```

Reply:
> This function is doing too much. You can tell as the doc is having to explain behavior that contradicts the name `findUpgradingNodes`.
> 
> Please change this to just return a list of upgrading nodes, the caller can then determine if the node is in the list, and handle that correct.

I went this way to avoid another loop to filter outside of the loop collecting the nodes with the annotation.

Reply:
> In fact this whole function can be removed if a label was used instead of an annotation. [Labels should be used when the use case is to find collections of objects that satisfy certain conditions](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#attaching-metadata-to-objects). Which is the case here.
> 
> That simplifies this to:
> 
> ```
> 	if err := c.List(ctx, nodeList, client.MatchingLabels{core.LabelOSStable: "windows", metadata.UpgradingAnnotation: "true"}); err != nil {
> 		return nil, fmt.Errorf("error listing Windows nodes: %w", err)
> 	}
> ```

Agree and it makes sense to me. I am unsure why/how the annotation ended up as the consented solution in past design sessions.

I will transition everything to a label; moving the filtering to the server side is a huge win here— @sebsoto, thanks for emphasizing this.